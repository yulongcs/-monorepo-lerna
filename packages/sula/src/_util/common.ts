import assignWith from 'lodash/assignWith';
import isUndefined from 'lodash/isUndefined';
import isArray from 'lodash/isArray';

function withDefined(dest: any, src: any) {
  return isUndefined(src) ? dest : src;
}

export function assignWithDefined(...args: Record<string, any>[]) {
  return assignWith(...args, withDefined);
}

export function toArray<T>(value?: T | T[] | null): T[] {
  if (value === undefined || value === null) {
    return [];
  }

  // @ts-ignore
  return isArray(value) ? value : [value];
}


