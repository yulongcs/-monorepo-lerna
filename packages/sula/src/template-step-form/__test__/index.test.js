import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { StepForm } from '../index';
import Memorize from '../memorize';
import { delay } from '../../__tests__/common';

const config = {
  mode: 'create',
  steps: [
    {
      title: 'Step1',
      subTitle: '',
      fields: [
        {
          name: 'input1',
          label: 'input1',
          field: 'input',
        },
      ],
    },
    {
      title: 'Step2',
      subTitle: '',
      fields: [
        {
          name: 'input2',
          label: 'input2',
          field: 'input',
        },
      ],
    },
    {
      title: 'Step3',
      subTitle: '',
      fields: [
        {
          name: 'input3',
          label: 'input3',
          field: 'input',
        },
      ],
    },
  ],
  submit: {
    url: '/success.json',
    method: 'post',
  },
};

async function buttonClick(wrapper, text, idx = 0) {
  wrapper
    .find('Memorize')
    .children()
    .childAt(idx)
    .find('button')
    .forEach((node) => {
      if (node.text() === text) {
        node.simulate('click');
      }
    });

  await act(async () => {
    await delay(1000);
    wrapper.update();
  });
}

describe('step form', () => {
  describe('step', () => {
    function getMemorizeItemStyle(wrapper, idx) {
      return wrapper.find('Memorize').children().childAt(idx).props().style;
    }

    function isActiveStep(wrapper, idx) {
      expect(
        wrapper.find('.ant-steps-item').at(idx).hasClass('ant-steps-item-active'),
      ).toBeTruthy();
    }

    it('step change', async () => {
      const wrapper = mount(<StepForm {...config} />);

      expect(getMemorizeItemStyle(wrapper, 0)).toEqual({});
      expect(getMemorizeItemStyle(wrapper, 1)).toEqual({ display: 'none' });
      expect(getMemorizeItemStyle(wrapper, 2)).toEqual({ display: 'none' });
      isActiveStep(wrapper, 0);

      await buttonClick(wrapper, 'Next');
      expect(getMemorizeItemStyle(wrapper, 0)).toEqual({ display: 'none' });
      expect(getMemorizeItemStyle(wrapper, 1)).toEqual({});
      isActiveStep(wrapper, 1);

      await buttonClick(wrapper, 'Next', 1);
      isActiveStep(wrapper, 2);

      expect(getMemorizeItemStyle(wrapper, 1)).toEqual({ display: 'none' });
      expect(getMemorizeItemStyle(wrapper, 2)).toEqual({});

      await buttonClick(wrapper, 'Previous', 2);
      isActiveStep(wrapper, 1);
      expect(getMemorizeItemStyle(wrapper, 0)).toEqual({ display: 'none' });
      expect(getMemorizeItemStyle(wrapper, 1)).toEqual({});
    });

    it('loading', async () => {
      const remoteValues = async () => {
        await delay(2000);
        return Promise.resolve({
          input1: 'a',
          input2: 'b',
        });
      };
      const wrapper = mount(<StepForm {...config} mode="edit" remoteValues={remoteValues} />);
      await delay(1000);
      wrapper.update();
      expect(wrapper.find('Spin').props().spinning).toEqual(true);

      await delay(2500);
      wrapper.update();
      expect(wrapper.find('Spin').props().spinning).toEqual(false);
    });

    it('result page', () => {
      const wrapper = mount(<StepForm {...config} result />);
      const memorizeItems = wrapper.find('Memorize').children();
      expect(memorizeItems.children().length).toEqual(4);
    });

    it('submit', async (done) => {
      const initialValues = {
        input1: 'a',
        input2: 'b',
      };
      const wrapper = mount(
        <StepForm
          {...config}
          submit={{
            url: '/submit.json',
            method: 'post',
            converter: ({ data }) => {
              expect(data).toEqual(initialValues);
              done();
            },
          }}
          initialValues={initialValues}
        />,
      );

      await buttonClick(wrapper, 'Next');
      await buttonClick(wrapper, 'Next', 1);
      await buttonClick(wrapper, 'Submit', 2);
    });

    it('deriction', () => {
      const wrapper = mount(<StepForm {...config} />);
      expect(wrapper.find('.sula-template-step-form-horizontal-steps').length).toBeTruthy();

      wrapper.setProps({ direction: 'vertical' });
      expect(wrapper.find('.sula-template-step-form-vertical-steps').length).toBeTruthy();
    });

    it('style', () => {
      const wrapper = mount(<StepForm {...config} />);
      wrapper.setProps({ stepsStyle: { fontSize: 16 } });
      expect(wrapper.find('.sula-template-step-form-horizontal-steps').props().style).toEqual({
        fontSize: 16,
      });

      wrapper.setProps({ formStyle: { fontSize: 16 } });
      expect(wrapper.find('.sula-template-step-form-horizontal-form').props().style).toEqual({
        fontSize: 16,
      });
    });
  });

  describe('memorize', () => {
    it('memorize', () => {
      expect(mount(<Memorize.Item />).html()).toBeNull();
    });
  });

  it('snapshot', () => {
    const wrapper = mount(<StepForm {...config} mode="view" />);
    expect(wrapper.render()).toMatchSnapshot();
  });
});
