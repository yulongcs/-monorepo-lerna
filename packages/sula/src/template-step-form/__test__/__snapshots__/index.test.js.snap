// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`step form snapshot 1`] = `
<div
  class="sula-template-step-form sula-template-step-form-horizontal"
>
  <div
    class="sula-template-step-form-horizontal-steps"
  >
    <div
      class="ant-steps ant-steps-horizontal ant-steps-default ant-steps-label-horizontal"
    >
      <div
        class="ant-steps-item ant-steps-item-process ant-steps-item-active"
      >
        <div
          class="ant-steps-item-container"
        >
          <div
            class="ant-steps-item-tail"
          />
          <div
            class="ant-steps-item-icon"
          >
            <span
              class="ant-steps-icon"
            >
              1
            </span>
          </div>
          <div
            class="ant-steps-item-content"
          >
            <div
              class="ant-steps-item-title"
            >
              Step1
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-steps-item ant-steps-item-wait"
      >
        <div
          class="ant-steps-item-container"
        >
          <div
            class="ant-steps-item-tail"
          />
          <div
            class="ant-steps-item-icon"
          >
            <span
              class="ant-steps-icon"
            >
              2
            </span>
          </div>
          <div
            class="ant-steps-item-content"
          >
            <div
              class="ant-steps-item-title"
            >
              Step2
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-steps-item ant-steps-item-wait"
      >
        <div
          class="ant-steps-item-container"
        >
          <div
            class="ant-steps-item-tail"
          />
          <div
            class="ant-steps-item-icon"
          >
            <span
              class="ant-steps-icon"
            >
              3
            </span>
          </div>
          <div
            class="ant-steps-item-content"
          >
            <div
              class="ant-steps-item-title"
            >
              Step3
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="sula-template-step-form-horizontal-form"
  >
    <div
      class="ant-spin-nested-loading"
    >
      <div
        class="ant-spin-container"
      >
        <form
          class="ant-form ant-form-horizontal"
        >
          <div>
            <div>
              <div
                class="ant-row ant-form-item"
              >
                <div
                  class="ant-col ant-col-8 ant-form-item-label"
                >
                  <label
                    class=""
                    for="input1"
                    title="input1"
                  >
                    input1
                  </label>
                </div>
                <div
                  class="ant-col ant-col-12 ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <input
                        class="ant-input ant-input-disabled"
                        disabled=""
                        id="input1"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="ant-row ant-form-item"
                style="margin-bottom: 0px;"
              >
                <div
                  class="ant-col ant-col-12 ant-col-offset-8 ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <div
                        class="sula-form-action-default"
                      >
                        <div
                          class="ant-space ant-space-horizontal ant-space-align-center"
                        >
                          <div
                            class="ant-space-item"
                            style="margin-right: 8px;"
                          >
                            <button
                              class="ant-btn ant-btn-primary"
                              type="button"
                            >
                              <span>
                                Next
                              </span>
                            </button>
                          </div>
                          <div
                            class="ant-space-item"
                          >
                            <button
                              class="ant-btn"
                              type="button"
                            >
                              <span>
                                Back
                              </span>
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              style="display: none;"
            >
              <div
                class="ant-row ant-form-item"
              >
                <div
                  class="ant-col ant-col-8 ant-form-item-label"
                >
                  <label
                    class=""
                    for="input2"
                    title="input2"
                  >
                    input2
                  </label>
                </div>
                <div
                  class="ant-col ant-col-12 ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <input
                        class="ant-input ant-input-disabled"
                        disabled=""
                        id="input2"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="ant-row ant-form-item"
                style="margin-bottom: 0px;"
              >
                <div
                  class="ant-col ant-col-12 ant-col-offset-8 ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <div
                        class="sula-form-action-default"
                      >
                        <div
                          class="ant-space ant-space-horizontal ant-space-align-center"
                        >
                          <div
                            class="ant-space-item"
                            style="margin-right: 8px;"
                          >
                            <button
                              class="ant-btn ant-btn-primary"
                              type="button"
                            >
                              <span>
                                Next
                              </span>
                            </button>
                          </div>
                          <div
                            class="ant-space-item"
                            style="margin-right: 8px;"
                          >
                            <button
                              class="ant-btn"
                              type="button"
                            >
                              <span>
                                Previous
                              </span>
                            </button>
                          </div>
                          <div
                            class="ant-space-item"
                          >
                            <button
                              class="ant-btn"
                              type="button"
                            >
                              <span>
                                Back
                              </span>
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              style="display: none;"
            >
              <div
                class="ant-row ant-form-item"
              >
                <div
                  class="ant-col ant-col-8 ant-form-item-label"
                >
                  <label
                    class=""
                    for="input3"
                    title="input3"
                  >
                    input3
                  </label>
                </div>
                <div
                  class="ant-col ant-col-12 ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <input
                        class="ant-input ant-input-disabled"
                        disabled=""
                        id="input3"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="ant-row ant-form-item"
                style="margin-bottom: 0px;"
              >
                <div
                  class="ant-col ant-col-12 ant-col-offset-8 ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <div
                        class="sula-form-action-default"
                      >
                        <div
                          class="ant-space ant-space-horizontal ant-space-align-center"
                        >
                          <div
                            class="ant-space-item"
                            style="margin-right: 8px;"
                          >
                            <button
                              class="ant-btn"
                              type="button"
                            >
                              <span>
                                Previous
                              </span>
                            </button>
                          </div>
                          <div
                            class="ant-space-item"
                          >
                            <button
                              class="ant-btn"
                              type="button"
                            >
                              <span>
                                Back
                              </span>
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
